README.md
# NAME

Mojolicious::Plugin::Gallery - Simple photo gallery for Mojolicious

# SYNOPSIS

    use Mojolicious::Plugin::Gallery;

# DESCRIPTION

Mojolicious::Plugin::Gallery use if you want to create simple gallery

Exampler content main.conf

    {
      gallery => {
        main_path => 'public/gallery',
        sizes => {
          thumbnail => {
            width   => 412,
            height  => 412,
            crop    => 1,
            quality => 85
          },
          medium => {
            width   => 800,
            height  => 550,
            crop    => 0,
            quality => 90
          },
          large => {
            width   => 1200,
            height  => 800,
            crop    => 0,
            quality => 90
          }
        }
      },
    }

example app.pl

    #!/usr/bin/env perl

    use lib 'lib';
    use Mojolicious::Lite;

    app->plugin(Config => { file => 'main.conf' });

    app->plugin('Gallery');

    get '/' => sub {
      my $c = shift;

      $c->render(text => '<a href="/gallery">Gallery</a>');
    };

    app->start;

Also you can find example app on https://github.com/sklukin/Mojolicious-Plugin-Gallery

Your steps for create new photo gallery

    - Make dir in public/gallery
    - Run ./cmd.pl resize
    - Update info in public/gallery/<your album>/meta.yml
    - Restart your app

# AUTHOR

sklukin <sklukin@yandex.ru>

# COPYRIGHT

Copyright 2020-2020 sklukin

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO


Changes
Revision history for Mojolicious-Plugin-Gallery

0.09  2020-09-16 01:50:21 MSK
        - Update pod

0.08  2020-09-16 01:50:21 MSK
        - Remove js and update plugin and test app

0.07  2020-09-16 01:50:21 MSK
        - Remove js and update plugin and test app

0.06  2020-09-15 01:50:21 MSK
        - Try push to cpan

0.05  2020-09-14 01:50:21 MSK
        - Add code of Plugin

0.04  2020-09-14 01:50:21 MSK
        - Add code of Plugin

0.03  2020-09-14 01:50:21 MSK
        - Add code of Plugin

0.02  2020-09-14 01:40:21 MSK
        - Update Readme

0.01  2020-09-14 01:38:21 MSK
        - Initial release


lib/Command/resize.pm
package Command::resize;
use Mojo::Base 'Mojolicious::Command';
use Mojo::File 'path';
use YAML::XS;
use DDP;

use File::Basename;
use GD;

use constant DEBUG => $ENV{PI_DEBUG} || 0;

has description => "Resize photos for gallery\n";
has usage       => "usage: $0\n";

sub run {
  my ($self) = @_;

  my $config = $self->app->config->{gallery};
  my $main_path = $config->{main_path};

  my $path = path $main_path;

  my $galleries = $path->list({dir => 1});
  for my $gallery ($galleries->each) {
    my $meta_source = $gallery->list->first(qr/yml/)->slurp;
    die 'meta is undef' unless $meta_source;
    my $meta = Load $meta_source;
    next if $meta->{skip};
    my $gallery_path = $gallery->to_string;

    my $photos = $gallery->list->grep(qr/jpg|png|JPG|PNG/);
    for my $photo ($photos->each) {
      my $file_path = $photo->to_string;

      _resize($file_path, $config->{sizes}, "$gallery_path");
      unlink $file_path;
      say "Done $file_path";
    }
  }
}

sub _resize {
  my ($file, $sizes, $imgDir) = @_;
  # https://github.com/davepagurek/Image-Resizer

  if (!$imgDir) {
    my ($a, $b, $c) = fileparse($file);
    $imgDir = $b;
  }

  if (!-d $imgDir) {
    mkdir $imgDir or die "Unable to create $imgDir";
  }

  if ($file && -e $file) {
    my $name = "img";
    my $mime = "jpg";
    if ($file =~ /[\/\\]*([a-zA-Z0-9-_ ]*)\.([a-z]+)$/i) {
      $name = $1;
      $mime = $2;
    }

    my $img;
    if (lc($mime) eq "jpg" || lc($mime) eq "jpeg") {
      $img = GD::Image->newFromJpeg($file);
    } elsif (lc($mime) eq "png") {
      $img = GD::Image->newFromPng($file);
    } else {
      die "Unsupported file format: $mime";
    }

    my ($w,$h) = $img->getBounds(); # find dimensions

    for my $size (keys %$sizes) {
      my $imgDirSize = "$imgDir/$size";
      if (!-d $imgDirSize) {
        mkdir $imgDirSize or die "Unable to create $imgDirSize";
      }
      my $newimg;
      if ($sizes->{$size}->{crop}) {
        my ($cut,$xcut,$ycut);
        if ($w>$h){
          $cut=$h;
          $xcut=(($w-$h)/2);
          $ycut=0;
        }
        if ($w<$h){
          $cut=$w;
          $xcut=0;
          $ycut=(($h-$w)/2);
        }
        $newimg = new GD::Image($sizes->{$size}->{width}, $sizes->{$size}->{height}, 1);
        $newimg->copyResampled($img,0,0,$xcut,$ycut,$sizes->{$size}->{width}, $sizes->{$size}->{height},$cut,$cut);
      } else {
        my $gd;
        if ($w>$h) {
          $newimg = new GD::Image($sizes->{$size}->{width}, (($h/$w)*$sizes->{$size}->{width}), 1);
          $newimg->copyResampled($img,0,0,0,0,$sizes->{$size}->{width}, (($h/$w)*$sizes->{$size}->{width}),$w,$h);
        } else {
          $newimg = new GD::Image(($w/$h)*$sizes->{$size}->{height}, ($sizes->{$size}->{height}), 1);
          $newimg->copyResampled($img,0,0,0,0,($w/$h)*$sizes->{$size}->{height}, ($sizes->{$size}->{height}),$w,$h);
        }
      }

      open(my $thumbFile, ">", "$imgDirSize/$name.jpg") or die "Cannot open $imgDirSize/$name.jpg: $!";
      binmode $thumbFile;
      print $thumbFile $newimg->jpeg($sizes->{$size}->{quality});
      close $thumbFile;
    }
  } else {
    die "Can't find file $file\n";
  }
}


1;


lib/Mojolicious/Plugin/Gallery.pm
package Mojolicious::Plugin::Gallery;

use strict;
use 5.008_005;
our $VERSION = '0.09';

use Mojo::Base 'Mojolicious::Plugin';
use Mojo::File 'path';
use File::Basename;
use YAML::XS;
use DDP;

sub register {
  my ($self, $app) = @_;
  my $config = $app->config->{gallery};

  my $path = path $config->{main_path};
  my $galleries = [];

  my $items = $path->list({dir => 1});
  for my $gallery ($items->each) {
    my $meta_source = $gallery->list->first(qr/yml|yaml/);
    die 'meta is undef' unless $meta_source;
    my $meta = Load $meta_source->slurp;
    next unless $meta->{show};

    my $gallery_name = $gallery->basename;
    my $gallery_path = $gallery->to_string;
    my $photos_path  = sprintf '%s/large', $gallery->to_string;

    my $photos = path($photos_path)->list->grep(qr/jpg|png/);
    my $text = '';
    my $photo_items;
    for my $photo ($photos->each) {
      my $filename = $photo->basename;
      my $gallery_path_withot_public = $gallery_path;

      # some Workaround. Fix it
      $gallery_path_withot_public =~ s/public\///;
      say $gallery_path_withot_public;

      push @$photo_items, {
        large     => sprintf('/%s/large/%s', $gallery_path_withot_public, $filename),
        medium    => sprintf('/%s/medium/%s', $gallery_path_withot_public, $filename),
        thumbnail => sprintf('/%s/thumbnail/%s', $gallery_path_withot_public, $filename),
      };
    }

    my $gallery_url = sprintf 'photos/%s', $meta->{url} || $gallery_name;

    my $c = $app->build_controller;
    $app->routes->get($gallery_url)->to(cb => sub {
      my $c = shift;
      $c->render(template => 'gallery_item',
        page_title   => $meta->{title},
        photos       => $photo_items,
        meta         => $meta,
      );
    });

    push @$galleries, {
      meta   => $meta,
      url    => $gallery_url,
      photos => [ splice(@$photo_items, 0, 3) ],
    };
  }

  $app->routes->get('/gallery')->to(cb => sub {
    my $c = shift;
    $c->render(template => 'gallery_list',
      page_title   => 'Somoe Photos',
      galleries    => $galleries,
    );
  });

}

1;
__END__

=encoding utf-8

=head1 NAME

Mojolicious::Plugin::Gallery - Simple photo gallery for Mojolicious

=head1 SYNOPSIS

  use Mojolicious::Plugin::Gallery;

=head1 DESCRIPTION

Mojolicious::Plugin::Gallery use if you want to create simple gallery

Exampler content main.conf

  {
    gallery => {
      main_path => 'public/gallery',
      sizes => {
        thumbnail => {
          width   => 412,
          height  => 412,
          crop    => 1,
          quality => 85
        },
        medium => {
          width   => 800,
          height  => 550,
          crop    => 0,
          quality => 90
        },
        large => {
          width   => 1200,
          height  => 800,
          crop    => 0,
          quality => 90
        }
      }
    },
  }


example app.pl

  #!/usr/bin/env perl

  use lib 'lib';
  use Mojolicious::Lite;

  app->plugin(Config => { file => 'main.conf' });

  app->plugin('Gallery');

  get '/' => sub {
    my $c = shift;

    $c->render(text => '<a href="/gallery">Gallery</a>');
  };

  app->start;

Also you can find example app on https://github.com/sklukin/Mojolicious-Plugin-Gallery

Your steps for create new photo gallery

  - Make dir in public/gallery
  - Run ./cmd.pl resize
  - Update info in public/gallery/<your album>/meta.yml
  - Restart your app

=head1 AUTHOR

sklukin E<lt>sklukin@yandex.ruE<gt>

=head1 COPYRIGHT

Copyright 2020-2020 sklukin

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
